### PR TITLE
Add support of Connection with FortiToken

### LINKED ISSUE #197: Connecting with FortiToken?
Hello, is there a way to connect to prompt for FortiToken? Thank you!

### PR DESCRIPTION
**Pull request by alagoutte**
Fix #197

### CODE REVIEW

**Comment by sloandroid** [comment]
I'm getting "Exception: C:\PowerFGT\PowerFGT\Public\Connection.ps1:181:17
Line |
181 | throw "Unable to connect to FortiGate""

Before I hadn't loaded the other functions and received

Show-FGTException: C:\PowerFGT\PowerFGT\Public\Connection.ps1:180:17
Line |
180 | Show-FGTException $_
| ~~~~~~~~~~~~~~~~~
| The term 'Show-FGTException' is not recognized as a name of a cmdlet, function, script file, or executable program.
| Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
InvalidOperation: C:\PowerFGT\PowerFGT\Public\Connection.ps1:185:21
Line |
185 | switch ($iwrResponse.Content[0]) {
| ~~~~~~~~~~~~~~~~~~~~~~~
| Cannot index into a null array.
Exception: C:\PowerFGT\PowerFGT\Public\Connection.ps1:216:17
Line |
216 | throw "Unable to found CSRF Cookie"
| ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| Unable to found CSRF Cookie

I'm using PowerShell 7.27 and firmware 6.4.9

**Comment by alagoutte** [comment]
What do you have when use get-module ? (how you have import the preview module ?)

can you try with -verbose ? (the Connect-FGT) and you need to use -token_code with your FortiToken

**Comment by sloandroid** [comment]
> 

I saved the updated module to my computer and imported it directly.

When I use
`Connect-FGT ipaddress -token_code tokenumber -verbose`

It says 
Verbsose: {
   username: my username
   secretkey: my secret key
   ajax: 1
}
Verbose: Post with 64-byte payload
Exception: C:\PowerFGT\Public\Connection.ps1:184
Line |
 184 |                  throw "Unable to connect to FortiGate"
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to connect to FortiGate

When using get-module it shows it as a Script 0.6.1

**Comment by alagoutte** [comment]
do you are using the port 443 for the management ?

**Comment by sloandroid** [comment]
> do you are using the port 443 for the management ?

I realized I use different port, but I changed it on my end. Now the only exception I get is:

Exception: \powerfgt\Public\Connection.ps1:233
Line |
 233 |                  throw "Unable to found CSRF Cookie"
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to found CSRF Cookie

I did have to use the -skipcertificatecheck, as I first received unable to connect (certificate) with a warning of "Invalid certicate (untrusted, wrong date, invalid name..

**Comment by alagoutte** [comment]
you can use -port XXX for the port

can you retry with -verbose ?

**Comment by alagoutte** [comment]
Ok Look good but something don’t work… I will try with 6.4.x

**Comment by alagoutte** [comment]
Ok, i need to add better check return error code when use token...

can you try to debug (using ssh) with the following command :
```
diagnose debug application httpd -1
diagnose debug enable
```

and try to connect using token

**Comment by sloandroid** [comment]
> Ok, i need to add better check return error code when use token...
> 
> can you try to debug (using ssh) with the following command :
> 
> ```
> diagnose debug application httpd -1
> diagnose debug enable
> ```
> 
> and try to connect using token

What's interesting is it does show the correct token in there even though it says it couldn't validate it. It never gets to prompting me for the 2FA.


 # [httpsd 20544 - 1668616803 critical] fweb_error_log[241] -- AH01964: Connection to child 0 established (server Fortigate:xxcorrectportxx)
[httpsd 20544 - 1668616803     info] fweb_debug_init[356] -- New POST request for "/logincheck" from "xx.xx.xx.xxx:12753"
[httpsd 20544 - 1668616803     info] fweb_debug_init[357] -- User-Agent: "Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.19044; en-US) PowerShell/7.3.0"
[httpsd 20544 - 1668616803     info] fweb_debug_init[359] -- Handler "logincheck-handler" assigned to request
[httpsd 20544 - 1668616803     info] logincheck_handler[413] -- entering vdom for login_attempt (vdom='root')
[httpsd 20544 - 1668616803     info] logincheck_handler[517] -- login attempt completed with code 2
[httpsd 20544 - 1668616803     info] logincheck_handler[523] -- login_attempt (method=5, vdom='root', name='xxxxx',admin_name='xxxx', auth_svr='')
[httpsd 20544 - 1668616803     info] logincheck_handler[532] -- 2FA -- token needed, orig_err=0
[httpsd 20544 - 1668616803    error] gui_send_mail_token_code[705] -- administrator does not have TFA enabled
[httpsd 20544 - 1668616803     info] output_response[58] -- sent response (status='3', buf='1')
[httpsd 20544 - 1668616803     info] fweb_debug_final[261] -- Completed POST request for "/logincheck" (HTTP 200)
[httpsd 20544 - 1668616803 critical] fweb_error_log[241] -- AH01991: SSL input filter read failed.
[httpsd 20544 - 1668616803 critical] fweb_error_log[241] -- AH01964: Connection to child 0 established (server Fortigate:xxcorrectportxx)
[httpsd 20544 - 1668616803     info] fweb_debug_init[356] -- New POST request for "/logincheck" from "xx.xx.xx.xxx:13265"
[httpsd 20544 - 1668616803     info] fweb_debug_init[357] -- User-Agent: "Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.19044; en-US) PowerShell/7.3.0"
[httpsd 20544 - 1668616803     info] fweb_debug_init[359] -- Handler "logincheck-handler" assigned to request
[httpsd 20544 - 1668616803     info] logincheck_handler[284] -- TFA token retrieved ('Correcttoken')
[httpsd 20544 - 1668616803     info] logincheck_handler[301] -- verifying token (vdom='root', admin='xxxx', token='xxxx', code='', chal_st='2', orig_err='0', profname='super_admin')
[httpsd 20544 - 1668616803    error] logincheck_handler[355] -- could not validate token (error=-111)
[httpsd 20544 - 1668616803     info] output_response[58] -- sent response (status='0', buf='')
[httpsd 20544 - 1668616803     info] fweb_debug_final[261] -- Completed POST request for "/logincheck" (HTTP 200)
[httpsd 20544 - 1668616803 critical] fweb_error_log[241] -- AH01991: SSL input filter read failed.

**Comment by alagoutte** [comment]
You are using -token_code xxxxxx with Xxxxx is the token id ?

**Comment by sloandroid** [comment]
I receive this now, and I've double checked the user name and password are correct

ERBOSE: {
  "username": "Correct User Name",
  "ajax": 1,
  "secretkey": "Password"
}
VERBOSE: HTTP/1.1 POST with 64-byte payload
VERBOSE: received 2-byte response of content type text/html
VERBOSE: {
  "secretkey": "Password",
  "ajax": 1,
  "token_code": "SerialNumberfortoken",
  "username": "Correct User Name"
}
VERBOSE: HTTP/1.1 POST with 92-byte payload
VERBOSE: received 1-byte response of content type text/html
Exception: C:\scripts\GithubDesktop\PowerFGT\PowerFGT\Public\Connection.ps1:210:21
Line |
 210 |  …             throw "Log in failure. Most likely an incorrect username/ …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Log in failure. Most likely an incorrect username/password combo

**Comment by sloandroid** [comment]
I figured out the issue for the last part. I was using my token serial, not my token from the MFA. Thank you! Module works great!

**Review by sloandroid — APPROVED** [review]
This worked for me.

File: PowerFGT/Public/Connection.ps1 [hunk 1]
```diff
@@ -173,6 +177,7 @@ function Connect-FGT {
             $postParams = @{username = $Credentials.username; secretkey = $Credentials.GetNetworkCredential().Password; ajax = 1 }
             $uri = $url + "logincheck"
             $iwrResponse = $null
+            Write-Verbose ($postParams | Convertto-json)
```
Comment (alagoutte) [inline PowerFGT/Public/Connection.ps1:180]: Remove verbose

**Review by alagoutte — COMMENTED** [review]
Missing .EXAMPLE and remove debug

**Pull request merged**